torchvision/ops/focal_loss.py

```diff
@@ -56,3 +57,95 @@ def sigmoid_focal_loss(
+                PyTorch's CrossEntropyLoss.
+                https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
+                The class dimension is expected to be absent, and each
+                element is the class value in the range [0, C).
+        alpha (Tensor[float]): Weighting factor in range (0,1) to balance
+                positive vs negative examples or None for no weighting. The elements
+                of this alpha should sum up to 1.0. Default: ``None``.
+        gamma (float): Exponent of the modulating factor (1 - p_t) to
+                balance easy vs hard examples. Default: ``2``.
+        eps (float): Small value to check if the sum of elements in alpha adds
+                up to 1.0.
+        reduction (string): ``'none'`` | ``'mean'`` | ``'sum'``
+                ``'none'``: No reduction will be applied to the output.
+                ``'mean'``: The output will be averaged.
+                ``'sum'``: The output will be summed.
+                ``'instance-sum-batch-mean'``: The output will be summed for each
+                        value in the batch, and then averaged across the entire
+                        batch. Default: ``'none'``.
+    Returns:
+        Loss tensor with the reduction option applied.
+    """
+    # Adapted from this version by Thomas V.
+    # https://discuss.pytorch.org/t/focal-loss-for-imbalanced-multi-class-classification-in-pytorch/61289/2
+    # Referenced from this github issue:
+    # https://github.com/pytorch/vision/issues/3250
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(softmax_focal_loss)
+
+    assert targets.dtype == torch.long, f"Expected a long tensor for 'targets', but got {targets.dtype}"
+
+    logits = inputs
+    weight = None
+    if alpha is not None:
+        num_classes = logits.size(1)
+        assert isinstance(alpha, torch.Tensor), f"Expected alpha to be torch.Tensor, got {type(alpha)}"
+        assert alpha.size(0) == num_classes, (
+            f"Expected alpha (weights) to have {num_classes} elements, but got {alpha.size(0)} elements"
+        )
+        assert abs(alpha.sum() - 1.0) <= eps, (
+            f"Expected elements of alpha to sum 1.0, instead they sum to {alpha.sum()}"
+        )
+        weight = alpha
+        
+    ce_loss = nn.functional.cross_entropy(logits, targets, reduction='none')
+    _ce_loss = nn.functional.cross_entropy(logits, targets, weight=weight, reduction='none')
+    # Instead of computing inputs.softmax(dim=1), we use the exponentiated
+    # negative log of the cross entropy loss.
+    #
+    # Why does this work?
+    # Since this is a multi-class setting, only one class is active. The
+    # probability of that class is 1, and the rest are all 0.
+    #
+    # Cross Entropy Loss computes:
+    # pt = softmax(...)
+    # loss = -1.0 * log(pt)
+    #
+    # Hence, exp(-loss) == pt
```

Comment:
Sadly I don't think this works anymore:

`loss = -weight_t * log(pt)`
`exp(-loss) == exp(weight_t * log(pt)`
`exp(-loss) == exp(log(pt ** weight_t))` due to the power rule of logarithms
`exp(-loss) == pt ** weight_t`


From what I've seen, I think this will instead require

```python
p = softmax(...)
ce_loss = nll_loss(p, weight=weight)
pt = p[:, targets]
```


Hence, 
`((1 - pt) ** gamma) * ce_loss` 
== `((1 - pt) ** gamma) * (-weight_t * log(pt))`
== `-weight_t * ((1 - pt) ** gamma) * log(pt)`

Reply:
@rehno-lindeque Please could you check the file changes and see if the code looks okay - I think the comment may suond misleading. The code is computing cross entropy loss twice, once without the weight and once with the weights, so that the one without can be exponentiated to compute the probability.

Reply:
Oh I did indeed miss that, you are right the code is functionally correct. Sorry about that.

I would still humbly suggest breaking down the cross_entropy into softmax followed by nll_loss as I described for performance reasons though.

That is, calculating cross entropy twice might be rather expensive, especially if the tensors are images. It's only necessary to calculate it once if you take `pt` directly from the softmax result.